\chapter{Exceptions and certified programming}

\section{Exceptions in programming languages}

``Exceptions'' is a rather broad term referring to a strategy of handling erroneous
states in computer programs by interrupting normal program execution, running special
code called an \emph{exception handler} and then resuming execution in a known, different
state.

The usual terminology is not very strict: the word ``exception'' may mean slightly
different things -- or different sides of the same thing -- in different contexts.
For example, programming languages are said to \emph{have exceptions} if they support
this kind of error handling; exceptions are said to \emph{be compiled}, while it is the
corresponding infrastructure and support code that is compiled; often the word ``exception''
denotes a piece of information about the error being handled; et cetera.

When an error occurs during normal execution of a program, \emph{an exception is thrown}%
\footnote{Some programming languages, such as Python or OCaml, use the term \emph{raised}
\cite{python:reference}.}.
This starts the
process of \emph{handling the exception}: looking for a suitable
\emph{exception handler} that \emph{handles the exception}, either by \emph{catching} it
to resume normal computation, or by \emph{re-throwing} it to find another handler
able to deal with the error.

As already mentioned, the word ``exception'' also denotes a piece of information about
the error or condition causing the exceptional state. Exceptions in this sense are
simply values of the programming language\footnote{In many languages, the choice of values that
can be thrown is restricted to those specialized for representing exceptions.}.
These values are provided by the code that throws,
handlers can inspect them and behave accordingly.

Due to common names of the corresponding syntactic features of popular programming languages%
\footnote{Most of them use the same keywords for this purpose.},
a piece of code together with attached pieces of handler code is called a \emph{try-block},
and a piece of handler code is called a \emph{catch-block}.

Most languages also provide \emph{finally-blocks}. These are pieces of code attached to a
try-block that are guaranteed to be executed after the try-block, whether an exception
has been thrown or not. Because of this property, finally-blocks are usually used to clean up
resources. In this thesis, we will not model finally-blocks as these can be
supplemented by an appropriate use of all-catching exception handlers and they are not
too useful in a language without side effects, anyway.

Often, there may be multiple handlers attached to a piece of code, each dealing with a different
kind of error. The appropriate handler
may be selected by the type of the exception being handled if the language is typed, with
most mainstream languages providing support for this approach, or by other means, for example
manual inspection of the value thrown.

A try-block needn't have handlers for all exceptions that might arise within. If an exception
is \emph{uncaught} within a try-block, it is \emph{propagated} to the containing try-block,
which may not catch this exception as well, propagating it further.
If an exception propagates all the way out of all nested try-blocks, the program usually
aborts.

To give a quick illustration how try-blocks look in the concrete syntax of some widely used
languages, \Fref{fig:try-blocks} contains four examples.\footnote{Note that OCaml does not
have syntax for finally-blocks; these are simulated by a function. Haskell does not have
syntax for exceptions at all, both \emph{catch} and \emph{finally} are just functions.
All code snippets are just symbolic and have been stripped of non-relevant context, such
as library imports and the definitions of the functions \emph{perform\_work}
and \emph{do\_cleanup}.}
%
\begin{figure}
\centering
\begin{subfigure}[b]{0.46\textwidth}\begin{codepy}
try:
	perform_work()
except IOError:
	print "IO error caught"
finally:
	do_cleanup()
\end{codepy}\caption{Python}\end{subfigure}
%
\begin{subfigure}[b]{0.46\textwidth}\begin{codejava}
try {
	performWork();
} catch (IOException e) {
	System.out.println(
	    "IO error caught");
} finally {
	doCleanup();
}
\end{codejava}\caption{Java}\end{subfigure}

\begin{subfigure}[b]{0.46\textwidth}\begin{codehs}
performWork
 `catch` (\(e :: IOException) ->
   putStrLn "IO error caught")
 `finally` 
   doCleanup
\end{codehs}\caption{Haskell}\end{subfigure}
%
\begin{subfigure}[b]{0.46\textwidth}\begin{codeml}
finally do_cleanup (fun () ->
  try perform_work ()
  with IO_error ->
    print_string "IO error caught"
) ()
\end{codeml}\caption{OCaml}\end{subfigure}

\caption{Try-blocks in different languages}
\label{fig:try-blocks}
\end{figure}

%\todo{Really include ``finally''? Revise OCaml.}

\subsection{Purpose}

As already mentioned, in practice, exceptions are mostly used to handle errors or
other exceptional states. The advantage to using
exceptions for this purpose is separation of concerns and hence cleaner resulting code.
Especially when reading a program, the reader first reads the code related the expected
execution path, uncluttered with error checks, which brings forward the main idea
of the code.

However, some languages, for example Python or OCaml, use exceptions also for control-flow
purposes, not only in rare and critical events. Python iterators raise an exception to
indicate the end of stream \cite{python:reference}, file-I/O functions in OCaml raise
an exception to indicate the end of file \cite{ocaml:reference}, OCaml programmers also
sometimes use exceptions to break loops early: for example, upon finding the element sought
in a list, etc. The implementation of
exceptions in these languages is efficient enough to make them cheap and enable this approach.

On the other hand, exceptions may also \emph{cause} subtle errors in programs. The code
is less obvious to read since there are no clues at which points the control flow may be diverted
because of an exception. Routines/functions have multiple exit points and these may
be non-obvious since exceptions may emerge from other parts of the code called within
the routine. In languages with manual memory management, resource leaks may occur.
Hence correctness of such code is more difficult to assess and reason about
by looking at the source code in isolation.

Still, the ubiquity of exceptions in practically all modern programming languages
hints at their great usefulness as the standard way of dealing with error states.

\subsection{History}

Exceptions have been present in programming languages for about half a century.
Louden and Lambert describe the invention of exceptions in their book
\emph{Programming Languages: Principles and Practice} as follows.

\begin{quote}
Exception handling was pioneered by the language PL/I in the 1960s and
significantly advanced by CLU in the 1970s, with the major design questions
eventually resolved in the 1980s and early 1990s. Today, virtually all major
current languages, including C++, Java, Ada, Python, ML, and Common Lisp (but
not C, Scheme or Smalltalk) have built-in exception handling mechanisms.
Exception handling has, in particular, been integrated very well into
object-oriented mechanisms in Python, Java, and C++, and into functional
mechanisms in ML and Common Lisp. Also, languages that do not have built-in
mechanisms sometimes have libraries available that provide them, or have other
built-in ways of simulating them. \cite[p.~423]{louden:languages}
\end{quote}

Over time, various flavors of exceptions and the corresponding infrastructure have evolved
in programming languages. Checked exceptions\footnote{A language with checked exceptions
requires its functions to specify in their types what exceptions they may throw.} (Java)
versus unchecked exceptions (Python) versus optionally checked exceptions (C++); some
languages restrict what kinds of values may be thrown as exceptions (Ruby), some do not (C++);
some languages do not provide finally-clauses (OCaml), some do not have \emph{any}
keywords designated for exceptions at all and use plain functions instead (Haskell), etc.

The variety is large but all these languages share the main principles of exception handling,
as described above.

\section{Certified programming}

The rise of dependently typed programming languages brought the opportunity to write
\emph{certified programs} in a quite convenient way. Such programs, written in languages
with very expressive type systems\footnote{Such as Agda, Coq or Epigram.}, can include
functions whose types represent theorems
about the programs themselves, usually about their adherence to specifications.
Then, under the Curry-Howard correspondence, implementations of these functions
represent proofs \cite{howard80}. These proofs are checked by typecheckers mechanically
without human intervention.

What's more, such programs needn't ensure their correctness only through pure theorems.
They usually also exploit
their type systems to assign far stricter types to their ordinary functions, blurring the line
between a type signature and a fully specified contract. Programs created in this way
can use the type system to enforce invariants in the program, providing a safety net
to catch programmer errors, and are often \emph{correct by construction}\footnote{In the
sense that types are set up to enforce the specification; an incorrect program would
not typecheck and subsequently compile.}.

The programming languages used for these purposes are usually \emph{total} functional
languages, meaning that all functions defined must be total. A total function terminates
on every possible input, always yielding a well-formed value. The trouble with non-termination
is that it makes the corresponding logic inconsistent -- and this is why totality is necessary
(not sufficient) for languages to be used in trustworthy developments.

Certification by a proof of correctness is fundamentally different from testing.
A proof gives a \emph{guarantee}, based on the source code of the program in question,
that the program follows its specification and never deviates. On the other hand, testing
is a stochastic
process that, although amenable to coverage improvements, is typically not exhaustible
and provides just a good \emph{chance} that if no defects are found, the program is correct.

\subsection{Certified compilers}

However, even when a program is certified, it does not exist in a vacuum. Leroy \cite{leroy09}
emphasizes that the program
itself is just a text -- but correctness of the program is assessed according to some specification,
the program needs to be turned into some low-level representation by a compiler and then it
needs to be executed on a machine in order to produce results. Hence, if the specification,
the compiler, the operating system, the machine, or any other component along the way
is incorrect, the actual result can no longer be guaranteed to be correct.

In fact, the impact of bugs in compilers and machines is much greater than that in ordinary
applications since all other software depends on them. Yet both compilers and processors
are very complex systems performing many non-transparent and complicated tasks.

One way of coping with these issues
is using compilers that are certified themselves.
For example, a practical compiler for quite
a broad subset of the C programming language, called CompCert \cite{leroy09} was developed
by Leroy et al. in the Coq proof assistant.

Hence, good specifications and certified compilers provably adhering to them would help
making programs more reliable wherever the gains of reliability outweigh the cost of
certification. As methods will improve, verification will get cheaper and cheaper; as
verification will cover more and more domains, software, compilers, operating systems,
hardware, etc., the scope of guarantees -- and subsequently gains -- will grow.

In this thesis, we will study a specific domain involving specification, compilation
and execution of programs with exceptions. We will specify
a semantics, create a compiler, prove its adherence to the specification, and examine
how this all works together practically in a dependently typed programming language.








































































